dash/_patch.py

```diff
@@ -0,0 +1,117 @@
+def _operation(name, location, **kwargs):
+    return {"operation": name, "location": location, "params": dict(**kwargs)}
+
+
+_noop = object()
+
+
+class Patch:
+    """
+    Patch a callback output value
+
+    Act like a proxy of the output prop value on the frontend.
+
+    Supported prop types: Dictionaries and lists.
+    """
+
+    def __init__(self, location=None, parent=None):
+        if location is not None:
+            self._location = location
+        else:
+            # pylint: disable=consider-using-ternary
+            self._location = (parent and parent._location) or []
+        if parent is not None:
+            self._operations = parent._operations
+        else:
+            self._operations = []
+
+    def __getitem__(self, item):
+        return Patch(location=self._location + [item], parent=self)
+
+    def __getattr__(self, item):
+        if item == "tolist":
+            # to_json fix
+            raise AttributeError
+        if item == "_location":
+            return self._location
+        if item == "_operations":
+            return self._operations
+        return self.__getitem__(item)
+
+    def __setattr__(self, key, value):
+        if key in ("_location", "_operations"):
+            self.__dict__[key] = value
+        else:
+            self.__setitem__(key, value)
+
+    def __delattr__(self, item):
+        self.__delitem__(item)
+
+    def __setitem__(self, key, value):
+        if value is _noop:
+            # The += set themselves.
+            return
+        self._operations.append(
+            _operation(
+                "Assign",
+                self._location + [key],
+                value=value,
+            )
+        )
+
+    def __delitem__(self, key):
+        self._operations.append(_operation("Delete", self._location + [key]))
+
+    def __add__(self, other):
+        self._operations.append(_operation("Add", self._location, value=other))
+        return _noop
+
+    def __iadd__(self, other):
+        self._operations.append(_operation("Add", self._location, value=other))
```

Comment:
Does this work for extending a list, ie `Patch() += [1,2,3]`?

Reply:
No this will trigger js undefined behavior, `[1,2] + [2,3] = "1,22,3"`, not sure there should really be multiple ways to do extend operations, maybe only accept numbers and strings for the numerical operations?

Reply:
My feeling is we should minimize how much you need to learn - and on the assumption that you already know what Python lists and dicts can do, that means we replicate the built-in behavior of lists and dicts as much as possible (including all the redundant built-in behavior like this), and wherever we can't or don't intend to replicate the built-in behavior we should be explicit about it.

So my first preference would be to just redirect to `extend` if `other` is a list (or tuple), but if you'd rather restrict to just strings and numbers please throw an error that directs users to `extend` if they pass a sequence.